File: Programação orientada a objetos/POO-241 - Trab 2 - Pedro_Quedevez/Veiculo.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include "StringUtils.hpp"
#include "Veiculo.hpp"
#include "Tokenizer.hpp"
#include "DateUtils.hpp"
#include "Publicacao.hpp"
#include "Excecao.hpp"

using namespace cpp_util;
using namespace std;

Veiculo::Veiculo() //
{
}

Veiculo::Veiculo(string csvline)
{

    Tokenizer tokens(csvline, ';');
    vector<string> campos = tokens.remaining();

    for (int i = 0; i < campos.size(); i++)
    {
        campos[i] = trim(campos[i]);
    }

    try
    {
        if (campos[2] == "P")
        {
            this->setSigla(campos[0]);
            if (siglasUtilizadas.find(campos[0]) != siglasUtilizadas.end())
            {
                throw Excecao("Código repetido para Veículo: " + campos[0] + ".");
            }
            this->setNome(campos[1]);
            this->setTipo(campos[2]);

            if (trim(campos[2]) != "P" && trim(campos[2]) != "C")
            {
                throw Excecao("Tipo de veículo desconhecido para veículo " + campos[0] + ": " + campos[2] + ".");
            }

            campos[3] = trocarVirgulaPorPonto(campos[3]);
            this->setFatorDeImpacto(stof(campos[3]));
            this->setIssn(campos[4]);
            this->publicacoes = vector<Publicacao>();
            siglasUtilizadas.insert(trim(campos[0]));
        }
        else if (campos[2] == "C")
        {
            this->setSigla(campos[0]);
            if (siglasUtilizadas.find(campos[0]) != siglasUtilizadas.end())
            {
                throw Excecao("Código repetido para Veículo: " + campos[0] + ".");
            }
            this->setNome(campos[1]);
            this->setTipo(campos[2]);

            if (trim(campos[2]) != "P" && trim(campos[2]) != "C")
            {
                throw Excecao("Tipo de veículo desconhecido para " + campos[0] + ": " + campos[2] + ".");
            }

            campos[3] = trocarVirgulaPorPonto(campos[3]);
            this->setFatorDeImpacto(stof(campos[3]));
            this->setIssn("");
            this->publicacoes = vector<Publicacao>();
            siglasUtilizadas.insert(trim(campos[0]));
        }
        else
        {
            throw Excecao("Tipo de veículo desconhecido para veículo " + campos[0] + ": " + campos[2] + ".");
        }
    }
    catch (Excecao e)
    {
        throw e;
    }
}

void Veiculo::addPublicacao(Publicacao p)
{
    this->publicacoes.push_back(p);
}

string Veiculo::getSigla()
{
    return this->sigla;
}
string Veiculo::getNome()
{
    return this->nome;
}
string Veiculo::getTipo()
{
    return this->tipo;
}
float Veiculo::getFatorDeImpacto()
{
    return this->fatorDeImpacto;
}
string Veiculo::getIssn()
{
    return this->issn;
}
void Veiculo::setSigla(string s)
{
    this->sigla = s;
}
void Veiculo::setNome(string s)
{
    this->nome = s;
}
void Veiculo::setTipo(string s)
{
    this->tipo = s;
}
void Veiculo::setFatorDeImpacto(float d)
{
    this->fatorDeImpacto = d;
}
void Veiculo::setIssn(string s)
{
    this->issn = s;
}

ostream &operator<<(ostream &c, Veiculo &d)
{
    stringstream fatorImpactoAux;
    fatorImpactoAux << fixed << setprecision(3) << d.getFatorDeImpacto();
    string fatorImpacto = trocarPontoPorVirgula(fatorImpactoAux.str());

    c << d.getSigla() << " " << d.getNome() << " " << d.getTipo() << " " << fatorImpacto << " " << d.getIssn();
    return c;
}

File: Programação orientada a objetos/POO 241 - Trab 1 - Pedro Quedevez/Veiculo.java
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.text.ParseException;
import java.util.Locale;
import java.util.ArrayList;

public class Veiculo {
    private String sigla;
    private String nome;
    private Character tipo;
    private Double fatorImpacto;
    private String issn;
    private ArrayList<Publicacao> publicacoes;

    static ArrayList<String> siglasCadastradas = new ArrayList<String>();

    public Veiculo(String csvline) throws excecao {
        String[] campos = csvline.split(";");
        DecimalFormatSymbols symbols = new DecimalFormatSymbols(Locale.getDefault());
        symbols.setDecimalSeparator(',');
        DecimalFormat decimalFormat = new DecimalFormat("#.##", symbols);
        try {
            if (campos.length == 5) {
                if (siglasCadastradas.contains(campos[0].trim())) {
                    throw new excecao("Codigo repetido para Veiculo: " + campos[0]);
                }
                siglasCadastradas.add(campos[0].trim());
                this.sigla = campos[0].trim();
                this.nome = campos[1].trim();
                if (campos[2].charAt(0) != 'C' && campos[2].charAt(0) != 'P') {
                    throw new excecao(
                            "Tipo de veículo desconhecido para veículo " + campos[0] + ": " + campos[2] + ".");
                }
                this.tipo = campos[2].charAt(0);
                try {
                    this.fatorImpacto = decimalFormat.parse(campos[3].trim()).doubleValue();
                } catch (ParseException e) {
                    e.printStackTrace();
                }
                this.issn = campos[4].trim();
                this.publicacoes = new ArrayList<Publicacao>();
            } else if (campos.length == 4) {
                if (siglasCadastradas.contains(campos[0].trim())) {
                    throw new excecao("Codigo repetido para Veiculo: " + campos[0].trim());
                }
                siglasCadastradas.add(campos[0].trim());
                this.sigla = campos[0].trim();
                this.nome = campos[1].trim();
                if (campos[2].charAt(0) != 'C' && campos[2].charAt(0) != 'P') {
                    throw new excecao(
                            "Tipo de veículo desconhecido para veículo " + campos[0].trim() + ": " + campos[2].trim()
                                    + ".");
                }
                this.tipo = campos[2].charAt(0);
                try {
                    this.fatorImpacto = decimalFormat.parse(campos[3].trim()).doubleValue();
                } catch (ParseException e) {
                    e.printStackTrace();
                }
                this.issn = "";
                this.publicacoes = new ArrayList<Publicacao>();
            }
        } catch (java.lang.NumberFormatException e) {
            System.out.println("Erro de formatacao");
        }
    }

    public static Character verificaTipoVeiculo(ArrayList<Veiculo> veiculos, String sigla) {
        for (Veiculo veiculo : veiculos) {
            if (veiculo.getSigla().trim().equals(sigla.trim())) {
                return veiculo.getTipo();
            }
        }
        return null;
    }

    @Override
    public String toString() {
        return sigla + ";" + nome + ";" + tipo + ";" + fatorImpacto + ";" + issn;
    }

    public String getSigla() {
        return sigla;
    }

    public void setSigla(String sigla) {
        this.sigla = sigla;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public Character getTipo() {
        return tipo;
    }

    public void setTipo(Character tipo) {
        this.tipo = tipo;
    }

    public Double getFatorImpacto() {
        return fatorImpacto;
    }

    public void setFatorImpacto(Double fatorImpacto) {
        this.fatorImpacto = fatorImpacto;
    }

    public String getIssn() {
        return issn;
    }

    public void setIssn(String issn) {
        this.issn = issn;
    }

    public ArrayList<Publicacao> getPublicacoes() {
        return publicacoes;
    }

    public void setPublicacoes(ArrayList<Publicacao> publicacoes) {
        this.publicacoes = publicacoes;
    }

    public void addPublicacao(Publicacao publicacao) {
        this.publicacoes.add(publicacao);
    }
}
